meson: add a new -Dstandalone-binaries=true option

filbranden opened the pull request:
cc @keszybz and @LorbusChris for [PR #25 in systemd Fedora RPM](https://src.fedoraproject.org/rpms/systemd/pull-request/25).

This adds an option to build standalone binaries that do not depend on the systemd shared libraries.

For now, only `systemd-sysusers` is also built as a standalone binary (but we're likely to add others such as tmpfiles soon.)

The standalone binaries are installed as `/usr/bin/%{name}.standalone`, the packaging system is reponsible for renaming those into the correct names during the packaging step. RPM is able to do so with `RemovePathPostfixes:`.

The default behavior is to build shared binaries only, since this option is mainly intended for building distribution packages.

Tested that a proper separate binary is built when using this option and that having it disabled (or using the default Meson configuration) does not produce a binary for this option.

Also tested building an RPM that packages the standalone packages, confirmed that the binaries inside it do not depend on systemd libraries and installing it conflicts with the main systemd package as expected.

LorbusChris commented:
Awesome, thank you for doing this!

filbranden commented:
And the corresponding changes to the specfile are in [PR #27  in Fedora Pagure](https://src.fedoraproject.org/rpms/systemd/pull-request/27).

poettering commented:
Hmm, so I take this this is because of the deps that libsystemd-shared.so pulls in, right? 

Which ones are those precisely?

I wonder if we couldn't just split libsystemd-shared instead, so that there's libsystemd-extra or so, which adds all the weirder deps?

I mean, I figure plenty of those deps are common enough so that even in your container images you have them anyway. But which deps are the other ones you hope to eliminate this way?

poettering commented:
This really could use some rationale in the commit msg

boucman commented:
could you add systemctl.standalone ? that would simplify cross-installation greatly since we need systemctl to set preset when building the image on a different architecture

currently we can build systemctl, but  `meson install` always install all targets and figuring out what is a dependency of what can be tricky....

filbranden commented:
@boucman :

> could you add `systemctl.standalone` ? that would simplify cross-installation greatly since we need `systemctl` to set preset when building the image on a different architecture.

Yes, this can be added for sure.

I would actually propose perhaps going further here, maybe split the `systemctl` code in two, one that manages enabling/disabling units and presets, that typically doesn't need to talk to the system manager, and another that does use D-Bus support.

That way we could build a `systemctl.standalone` that only works for the operations it will actually be able to perform in such a system and clearly error out on the others.

In any case, this `*.standalone` scheme opens the doors to such possibilities too.

(And straight building a `systemctl.standalone` with no other modifications would also work. I built a `systemd-tmpfiles` in a follow up commit and it was quite trivial to do.)

poettering commented:
I am not suggesting splitting the library up into one library for each dep. I am just saying maybe we can split it in two: the obvious stuff that always has to be there, and everything else.

I mean, there's no point I think in splitting out code hat links against libselinux, since I doubt it's realistic to have a Fedora container without it anyway. OTOH libp11-kit is something that a container might very well do without, so it might be in the second, split out library instead.

That all said, in the TODO list we have one item that will make this excercise even more complex: it was my intention to add an --image= switch in most our tools. i.e. that in addition to --root= where they operate relative to some OS tree they can directly operate on a disk image. i.e. the idea would be to extend what nspawn supports to all our tools: journalctl --image= pulls logs out of a disk image, systemd-sysusers --image= generates the users inside of a disk image, bootctl --image= can install a boot loader directly onto a disk image, and so on. For all our tools. But that means that libcryptsetup and libblkid would become direct deps from most of our tools, and such a standalone thing wouldn't work anymore... And the latte rpulls in libgcrypt iirc and so on.

But again, my question was about which deps you actually care about you want to get rid of. i mean, instead of the readelf dump above librt, libc, libpthread, libcrypt are nothing you could sensibly get rid of anyway. So which ones are the libs you actually *can* get rid of? i.e. where it's realistic that for minimal container images you don't pull them in? i.e. which are the deps only very few packages in your basic OS image currently use, where it is worth doing the work of making them optional?

I see at least two approaches to reach your goal besides the one you proposed: 

1. figure out if we can reasonably do dlopen() on some of the deps. For example llibcryptsetup might actually be a candidate for that. the API is not too big. We are already doing that for libxkbcommon for example, doing that for one or two more "leaf" libraries with limited APIs might be doable.

2. split libshared into two: one that has all the deps in them that you cannot reasonably get rid of anyway, and one other that has the rest

But before we choose a path here, let's figure out why you actually want to do this? i.e. what do you want to cut out? which of the deps in the readelf output you are going after?

filbranden commented:
@LorbusChris Can you answer Lennart's question about the motivations and which dependencies you would prefer to have removed from the standalone binaries?

LorbusChris commented:
My goal here was: "I want to be able to install only {sysusers,tmpfiles} without dragging in all of systemd"

FWIW for Fedora Containers I'd be content with what I was trying to do in https://src.fedoraproject.org/rpms/systemd/pull-request/25 -- i.e. just the ability to distribute the sysusers and tmpfiles binaries separately, even if that means they'll still depend on, and pull in, the libsystemd-shared object.

A drawback of that is of course the bigger size compared to this approach here,
OTOH there wouldn't be conflicting shared vs standalone packages - I'm somewhat weary of that.

It sounds like both of the approaches @poettering suggested could be combined? -- even if they won't take us all the way to the size savings we get from this here, both sound very sane to me.

filbranden commented:
Also note that right now the standalone binary doesn't pull libgcrypt, libcryptsetup, libp11-kit, etc. Only libselinux is included (and an indirect dependency on pcre2 apparently.)

```
$ readelf -d build/systemd-sysusers.standalone

Dynamic section at offset 0xa93a0 contains 31 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libselinux.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```

And:

```
$ ldd build/systemd-sysusers.standalone
	linux-vdso.so.1 (0x00007ffc8dd82000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f6a2a793000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f6a2a766000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f6a2a744000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f6a2a57a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f6a2a7be000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007f6a2a4e1000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f6a2a4da000)
```

I imagine Meson is doing an "as needed" link here. So in a way this *also* solves the problem with the dependency libraries we might not want to pull in on every container...

jamescassell commented:
What's the space (and memory?) savings of systemd-shared.so versus linking each component separately? I'd prefer having the primary copy of separately useful tools being linked separately from the rest, versus having 2 different versions that could be installed.

poettering commented:
let me ask this differently then: what is the actual size difference between two OS images:

1. one where libsystemd-shared is its own RPM, but otherwise linked like it currently is, and you only include systemd-tmpfiles + systemd-sysusers in the image, thus only pulling in this new libsystemd-shared dep and its transitive deps, but not the systemd main binary, i.e. no pid1 or so.

2. one where you do what this PR does comprehensively, i.e. build both systemd-tmpfiles and systemd-sysusers with libsystemd-shared built into it, without dynamic linking for that. (in this case you'll ship most of the libsystemd-shared code twice of course!)

Unless we have comparable data for these two cases I don't think we can make an educated decision on this.

keszybz commented on meson.build:
```diff
@@ -2930,6 +2930,21 @@ if conf.get('ENABLE_SYSUSERS') == 1
+                                     libsystemd_static,
+                                     libjournal_client],
+                        install_rpath : rootlibexecdir,
+                        install : true,
```
It'd be better to move the conditional here, i.e. only build and install the binary if configured, but allow building on demand always. Something like
```python
           build_by_default: have_standalone_binaries,
           install : have_standalone_binaries,
```

filbranden replied:
Good call!

I also noticed that `install_rpath` is not needed here, since it was actually needed for the `systemd-shared` solib, which is no longer present.

Tested the `have_standalone_binaries` change and it works as expected for me.

filbranden commented:
I applied the suggestions from @keszybz and also included a second commit on the stack to also build `systemd-tmpfiles` as a standalone binary.

(I'm happy to do a follow up on a standalone `systemctl` but I think that one would benefit from some code changes that would allow it to enable/disable/preset but not do any D-Bus calls, since I'd expect that's the purpose of having one.)

Also improved the description of the first commit (the one introducing `-Dstandalone-binaries=true`) with a longer description of the motivation for it and also mentioning the container use case.

Yes, we should definitely update CI to enable this option to have some test coverage on building and linking these binaries. Who can take care of doing so on the CentOS CI? I guess we should do that on Travis-CI as well, right? I can prepare a commit for the latter...

Cheers!
Filipe

keszybz commented:
Ooops:
```
524/802 check-help-systemd-sysusers.standalone  FAIL     0.04 s (exit status 2)

--- command ---
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-sysusers.standalone
--- stdout ---
systemd-sysusers.standalone --help output is empty.
--- stderr ---
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: 7: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-sysusers.standalone: not found
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: 14: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-sysusers.standalone: not found
-------

526/802 check-help-systemd-tmpfiles.standalone  FAIL     0.04 s (exit status 2)

--- command ---
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-tmpfiles.standalone
--- stdout ---
systemd-tmpfiles.standalone --help output is empty.
--- stderr ---
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: 7: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-tmpfiles.standalone: not found
/tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: 14: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/tools/check-help.sh: /tmp/autopkgtest.P1C5fO/build.dVk/systemd/build-deb/systemd-tmpfiles.standalone: not found
```
And this is actually caused by my suggested change :(
One option would be to undo that change.
Second option would be to only add the executable() to public_programs if the option is set.
Third option would be to make the help output test depend on the binary.

keszybz commented on meson.build:
```diff
@@ -2952,6 +2967,22 @@ if conf.get('ENABLE_TMPFILES') == 1
+                'systemd-tmpfiles.standalone',
+                'src/tmpfiles/tmpfiles.c',
+                'src/tmpfiles/offline-passwd.c',
+                'src/tmpfiles/offline-passwd.h',
```
I think it'd be better to define a `systemd_tmpfiles_sources` in `src/tmpfiles/meson.build` and reference it here and above, rather than to repeat this list.

filbranden replied:
Done!

keszybz commented:
> One option would be to undo that change.

Maybe that the best option? I'm trying to use this more flexible scheme in new projects, but systemd was done in the old style with outer conditionals. We should do a similar change comprehensively to the whole meson.spec file, so maybe it's better to just do this one in the old style too.

filbranden commented:
> > One option would be to undo that change.
>
> Maybe that the best option?

Yeah, I briefly experimented with `if want_tests != 'false' and (not name.endswith('.standalone') or have_standalone_binaries)` in the check-help test definition, but that feels thoroughly unsatisfying since it's hardcoding these two assumptions about the binary extension and the config option...

Reverted to only update `public_programs` to add the new binaries when the option is set.

(Agree that it's a cool way to make it conditional! I really liked the initial suggestion.)

Also addressed the bit about tmpfiles sources. PTAL.

filbranden commented:
Should I update `travis-ci/managers/{fedora,debian}.sh` to include `-Dstandalone-binaries=true` in RUN and RUN_CLANG? Happy to push an additional commit to this same PR... Or we can address this in a follow up PR, whatever you prefer.

Or do you think another one of the CI systems should take this option instead?

jamescassell commented:
Would it make sense to create a new systemd-core.so that includes only the things needed for tmpfiles, sysusers, and systemctl, then have systemd-shared.so link against this new `.so`? (Or everything that doesn't bring external deps into the new -core.so?)

keszybz commented:
> let me ask this differently then: what is the actual size difference between two OS images:

Let's approximate:
1. `dnf install glibc libselinux` and run `du` → 194M
2. `dnf install systemd` and remove systemd and all depend packages and run `du` → 430M

To 1. we have to add systemd-sysusers.static, ~200kb, the end result is 194.2 MB.
To 2. we have to add systemd-shared.so + systemd-sysuers, 2.7 MB + ~60kb, the end result is 431 MB.

I think this answers any questions.

keszybz commented:
> Would it make sense to create a new systemd-core.so that includes only the things needed for tmpfiles, sysusers, and systemctl, then have systemd-shared.so link against this new `.so`? (Or everything that doesn't bring external deps into the new -core.so?)

The proof is in the pudding, as they say. My guess would be that this would be very hard to do in a way that creates significant savings, because the code in systemd-shared is very finely intertwined. But I may be wrong. To answer the question, it would be necessary to actually do the work and create the split.

jamescassell commented:
> > let me ask this differently then: what is the actual size difference between two OS images:
> 
> Let's approximate:
> 1. `dnf install glibc libselinux` and run `du` → 194M
> 2. `dnf install systemd` and remove systemd and all depend packages and run `du` → 430M
> 
> To 1. we have to add systemd-sysusers.static, ~200kb, the end result is 194.2 MB.
> To 2. we have to add systemd-shared.so + systemd-sysuers, 2.7 MB + ~60kb, the end result is 433 MB.
> 
> I think this answers any questions.
> 

Please also use the glibc-minimal-langpack and coreutils-single packages when doing the comparisons, to trim down dependencies further. I still think it's best to split systemd-shared.so into parts with external deps and parts without.

keszybz commented:
> Please also use the glibc-minimal-langpack and coreutils-single packages

I used glibc-minimal-langpack, but omitted if from the listing for simplification. But I don't think this changes much: g-m-l and c-s mostly create an offset that is same in both cases. And when not same, the benefit would be bigger in the .standalone approach more (since current systemd rpm depends on more things, so it gains less when dependencies are split).

poettering commented:
> I think this is reasonable. The standalone version is ~200K and only has libc and libselinux as deps, which are pretty much guaranteed to be installed on any selinux system. The normal version is a bit smaller but pulls in a tree of deps, including two crypto stacks and libudev and a bunch of other hardware-related libraries.

Well, but are this two crypto stacks really avoidable? I mean, I am the strong suspicion that the most minimal Fedora container will contain openssl one way or another. If we don't pull it in, then something else will.

I doubt that openssl is any more redundant than glibc or libselinux is.

> All of that might not be necessary in a small container. This means that even if we manage to somehow split systemd-shared (which I think it going to be hard, but let's ignore that for now), it will not give comparable savings.

I have my doubts you can really avoid that stuff.

i.e. if you build a minimal fedora container right now, which are the libs that libsystemd-shared pulls in that nothing else pulls in? Or which libs does libsystemd-shared pull in that only "very few" other packages in the minimal set pull in, and are candidates to be kicked out?

unless that question is answered everything here is premature optimization

poettering commented:
> > let me ask this differently then: what is the actual size difference between two OS images:
> 
> Let's approximate:
> 
>     1. `dnf install glibc libselinux` and run `du` → 194M
> 
>     2. `dnf install systemd` and remove systemd and all depend packages and run `du` → 430M
> 
> 
> To 1. we have to add systemd-sysusers.static, ~200kb, the end result is 194.2 MB.
> To 2. we have to add systemd-shared.so + systemd-sysuers, 2.7 MB + ~60kb, the end result is 433 MB.
> 
> I think this answers any questions.

No it does not. It doesn't tell us anything about what precisely is the delta of 236M. And yes, if you install all of systemd it's gonna grow, no doubt. But let's no be sloppy here. let's do our homework first and figure what's exactly pulled in here and why and what we can get rid of.

i.e. how large is an image that starts out by your image from point 1, and then gets copied in just the plain systemd-tmpfiles binary and all files it strictly needs to run right now?

poettering commented:
or to ask the question differently: if you tell dracut to build an initrd consisting only of the systemd-tmpfiles binary, how large does the image actually get?

poettering commented:
i mean, it's fine doing such hacks downstream without the necessary research, if you want quick results. I totally get that. But for something to be merged upstream we need the research first, and do it properly. upstream is not the place for sloppy hacks, sorry.

poettering commented:
> But before we choose a path here, let's figure out why you actually want to do this? i.e. what do you want to cut out? which of the deps in the readelf output you are going after?
>
> @LorbusChris Can you answer Lennart's question about the motivations and which dependencies you would prefer to have removed from the standalone binaries?

The question regarding the why may be answered. May questions regarding the deps this is precisely about are still unanswered.

boucman commented:
Well, fedora is not exactly "minimal"... The whole container ecosystem is really not what you want to base your thought process on when thinking of minimal dependencies. These systems are based on pre-compiled distros (usually debian or fedora) and most packages are compiled with various optional dependencies to match multiple use-cases.

On embedded systems we custom-compile everything (mainly for traçability purposes, but size is very important too.) This means that we end up with far less dependencies for minimal systems and that systems with no network access (debug ver serial line, for example) will not have any SSL library at all.

I can build a minimal buildroot-based image for you, but a minimal buildroot image with systemd is bascially systemd+busybox and nothing else. So any dependency that is not pulled in by systemd won't be pulled in at all.

For the cross-compile use-case it's even worse. If we make those two ssl stacks mandatory, it means that we need to compile all of them for the host and the target every time we build systemd. systemd's compile time is not neglectable when compiling a whole system, and compiling it twice is already not funny. So anything that allows us to minimize that would be nice.

topimiettinen commented:
> or to ask the question differently: if you tell dracut to build an initrd consisting only of the systemd-tmpfiles binary, how large does the image actually get?

This is not made by dracut, but handcrafted (using ldd for libraries). Probably more files in /etc/selinux could be pruned but files aren't so big so I haven't done it yet. Most files are from Debian but I've recompiled a few packages without optimizations so the file sizes may be bigger than normal.
```
# du -hs /usr/local/lib/rootimages/systemd-tmpfiles-all.squashfs 
8.7M    /usr/local/lib/rootimages/systemd-tmpfiles-all.squashfs
# unsquashfs -ll /usr/local/lib/rootimages/systemd-tmpfiles-all.squashfs 
Parallel unsquashfs: Using 4 processors
98 inodes (280 blocks) to write

drwxr-xr-x root/root               139 2020-06-08 19:29 squashfs-root
lrwxrwxrwx root/root                 7 2019-03-12 19:29 squashfs-root/bin -> usr/bin
drwxr-xr-x root/root                 3 2020-06-08 19:29 squashfs-root/dev
drwxr-xr-x root/root                96 2020-06-08 19:29 squashfs-root/etc
-rw-r--r-- root/root              1279 2020-02-16 12:11 squashfs-root/etc/group
-rw-r--r-- root/root               536 2019-12-11 09:41 squashfs-root/etc/nsswitch.conf
-rw-r--r-- root/root              3274 2020-04-12 22:37 squashfs-root/etc/passwd
drwxr-xr-x root/root                44 2020-06-08 19:29 squashfs-root/etc/selinux
-rw-r--r-- root/root               603 2020-02-15 12:26 squashfs-root/etc/selinux/config
drwxr-xr-x root/root                60 2020-06-08 19:29 squashfs-root/etc/selinux/default
drwxr-xr-x root/root               396 2020-06-08 19:29 squashfs-root/etc/selinux/default/contexts
-rw-r--r-- root/root                 0 2020-05-02 12:20 squashfs-root/etc/selinux/default/contexts/customizable_types
-rw-r--r-- root/root               195 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/dbus_contexts
-rw-r--r-- root/root              1170 2020-02-16 09:59 squashfs-root/etc/selinux/default/contexts/default_contexts
-rw-r--r-- root/root               114 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/default_type
-rw-r--r-- root/root                41 2020-02-16 10:11 squashfs-root/etc/selinux/default/contexts/failsafe_context
drwxr-xr-x root/root               253 2020-06-08 19:29 squashfs-root/etc/selinux/default/contexts/files
-rw-r--r-- root/root            153227 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts
-rw-r--r-- root/root           3521964 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts.bin
-rw-r--r-- root/root             28488 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts.homedirs
-rw-r--r-- root/root            624100 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts.homedirs.bin
-rw-r--r-- root/root              1270 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts.local
-rw-r--r-- root/root             25272 2020-06-05 14:35 squashfs-root/etc/selinux/default/contexts/files/file_contexts.local.bin
-rw-r--r-- root/root                37 2020-05-04 11:51 squashfs-root/etc/selinux/default/contexts/files/file_contexts.subs
-rw-r--r-- root/root               896 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/files/file_contexts.subs_dist
-rw-r--r-- root/root               139 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/files/media
-rw-r--r-- root/root                30 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/initrc_context
-rw-r--r-- root/root               144 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/lxc_contexts
-rw-r--r-- root/root                20 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/openrc_contexts
-rw-r--r-- root/root                33 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/removable_context
-rw-r--r-- root/root                18 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/securetty_types
-rw-r--r-- root/root              1170 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/sepgsql_contexts
-rw-r--r-- root/root                30 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/userhelper_context
drwxr-xr-x root/root               107 2020-06-08 19:29 squashfs-root/etc/selinux/default/contexts/users
-rw-r--r-- root/root               253 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/users/guest_u
-rw-r--r-- root/root               871 2020-02-16 10:01 squashfs-root/etc/selinux/default/contexts/users/root
-rw-r--r-- root/root               618 2020-02-15 13:24 squashfs-root/etc/selinux/default/contexts/users/staff_u
-rw-r--r-- root/root               552 2020-05-02 12:20 squashfs-root/etc/selinux/default/contexts/users/unconfined_u
-rw-r--r-- root/root               334 2020-02-15 14:01 squashfs-root/etc/selinux/default/contexts/users/user_u
-rw-r--r-- root/root               307 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/users/xguest_u
-rw-r--r-- root/root                29 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/virtual_domain_context
-rw-r--r-- root/root                71 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/virtual_image_context
-rw-r--r-- root/root              2920 2019-02-21 15:09 squashfs-root/etc/selinux/default/contexts/x_contexts
drwxr-xr-x root/root                32 2020-06-08 19:29 squashfs-root/etc/selinux/default/policy
-rw-r--r-- root/root            727170 2020-06-05 14:35 squashfs-root/etc/selinux/default/policy/policy.32
-rw-r--r-- root/root               147 2020-06-05 14:35 squashfs-root/etc/selinux/default/seusers
drwxr-xr-x root/root               139 2020-06-08 19:29 squashfs-root/etc/tmpfiles.d
-rw-r--r-- root/root               777 2020-04-14 15:02 squashfs-root/etc/tmpfiles.d/local-dev.conf
-rw-r--r-- root/root               176 2016-09-17 13:53 squashfs-root/etc/tmpfiles.d/local-nvidia-kernel.conf
-rw-r--r-- root/root              1605 2020-02-24 18:05 squashfs-root/etc/tmpfiles.d/local-proc.conf
-rw-r--r-- root/root              1454 2019-12-08 15:04 squashfs-root/etc/tmpfiles.d/local-sys.conf
-rw-r--r-- root/root               808 2020-02-04 12:45 squashfs-root/etc/tmpfiles.d/static-nodes.conf
lrwxrwxrwx root/root                 7 2019-03-12 19:29 squashfs-root/lib -> usr/lib
lrwxrwxrwx root/root                 9 2019-03-12 19:29 squashfs-root/lib64 -> usr/lib64
drwxr-xr-x root/root                 3 2020-06-08 19:29 squashfs-root/proc
drwxr-xr-x root/root                 3 2020-06-08 19:29 squashfs-root/run
drwxr-xr-x root/root                 3 2020-06-08 19:29 squashfs-root/sys
drwxr-xr-x root/root                 3 2020-06-08 19:29 squashfs-root/tmp
drwxr-xr-x root/root                50 2020-06-08 19:29 squashfs-root/usr
drwxr-xr-x root/root                39 2020-06-08 19:29 squashfs-root/usr/bin
-rwxr-xr-x root/root            210232 2020-04-27 18:38 squashfs-root/usr/bin/systemd-tmpfiles
drwxr-xr-x root/root                72 2020-06-08 19:29 squashfs-root/usr/lib
drwxr-xr-x root/root                47 2020-06-08 19:29 squashfs-root/usr/lib/systemd
-rw-r--r-- root/root           7403576 2020-04-27 18:38 squashfs-root/usr/lib/systemd/libsystemd-shared-245.so
drwxr-xr-x root/root               387 2020-06-08 19:29 squashfs-root/usr/lib/tmpfiles.d
-rw-r--r-- root/root                35 2020-05-06 17:22 squashfs-root/usr/lib/tmpfiles.d/cryptsetup.conf
-rw-r--r-- root/root               365 2020-06-02 21:48 squashfs-root/usr/lib/tmpfiles.d/dbus.conf
-rw-r--r-- root/root               517 2020-03-13 16:45 squashfs-root/usr/lib/tmpfiles.d/debian.conf
-rw-r--r-- root/root               362 2019-07-21 09:23 squashfs-root/usr/lib/tmpfiles.d/home.conf
-rw-r--r-- root/root              1096 2019-07-21 09:23 squashfs-root/usr/lib/tmpfiles.d/journal-nocow.conf
-rw-r--r-- root/root               841 2020-04-23 12:46 squashfs-root/usr/lib/tmpfiles.d/legacy.conf
-rw-r--r-- root/root                32 2020-05-26 23:51 squashfs-root/usr/lib/tmpfiles.d/libreswan.conf
-rw-r--r-- root/root                61 2020-03-03 14:21 squashfs-root/usr/lib/tmpfiles.d/lvm2.conf
-rw-r--r-- root/root                33 2020-06-01 20:17 squashfs-root/usr/lib/tmpfiles.d/man-db.conf
-rw-r--r-- root/root               239 2020-02-07 16:54 squashfs-root/usr/lib/tmpfiles.d/passwd.conf
-rw-r--r-- root/root               697 2020-05-17 20:28 squashfs-root/usr/lib/tmpfiles.d/static-nodes-permissions.conf
-rw-r--r-- root/root               305 2020-05-14 03:34 squashfs-root/usr/lib/tmpfiles.d/sudo.conf
-rw-r--r-- root/root               597 2020-04-12 16:50 squashfs-root/usr/lib/tmpfiles.d/systemd-nologin.conf
-rw-r--r-- root/root               823 2020-04-12 16:50 squashfs-root/usr/lib/tmpfiles.d/systemd-tmp.conf
-rw-r--r-- root/root              1792 2020-04-27 18:38 squashfs-root/usr/lib/tmpfiles.d/systemd.conf
-rw-r--r-- root/root               448 2020-04-27 18:38 squashfs-root/usr/lib/tmpfiles.d/tmp.conf
-rw-r--r-- root/root               568 2020-04-27 18:38 squashfs-root/usr/lib/tmpfiles.d/var.conf
-rw-r--r-- root/root               655 2019-07-21 09:23 squashfs-root/usr/lib/tmpfiles.d/x11.conf
drwxr-xr-x root/root               703 2020-06-08 19:29 squashfs-root/usr/lib/x86_64-linux-gnu
-rw-r--r-- root/root             38936 2020-05-01 06:41 squashfs-root/usr/lib/x86_64-linux-gnu/libacl.so.1
-rw-r--r-- root/root             34904 2019-01-13 15:20 squashfs-root/usr/lib/x86_64-linux-gnu/libargon2.so.1
-rw-r--r-- root/root            128944 2020-04-12 18:44 squashfs-root/usr/lib/x86_64-linux-gnu/libaudit.so.1
-rw-r--r-- root/root            326624 2020-05-22 13:41 squashfs-root/usr/lib/x86_64-linux-gnu/libblkid.so.1
-rw-r--r-- root/root           1831600 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/libc.so.6
-rw-r--r-- root/root             26976 2020-05-19 15:54 squashfs-root/usr/lib/x86_64-linux-gnu/libcap-ng.so.0
-rw-r--r-- root/root             43536 2020-05-16 11:20 squashfs-root/usr/lib/x86_64-linux-gnu/libcap.so.2
-rw-r--r-- root/root            202680 2020-04-05 02:12 squashfs-root/usr/lib/x86_64-linux-gnu/libcrypt.so.1
-rw-r--r-- root/root           3044192 2020-04-21 22:45 squashfs-root/usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
-rw-r--r-- root/root            478608 2020-05-06 17:22 squashfs-root/usr/lib/x86_64-linux-gnu/libcryptsetup.so.12
-rw-r--r-- root/root            432688 2020-03-03 14:21 squashfs-root/usr/lib/x86_64-linux-gnu/libdevmapper.so.1.02.1
-rw-r--r-- root/root             14592 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/libdl.so.2
-rw-r--r-- root/root           1163960 2020-02-23 14:39 squashfs-root/usr/lib/x86_64-linux-gnu/libgcrypt.so.20
-rw-r--r-- root/root            137424 2020-02-13 01:14 squashfs-root/usr/lib/x86_64-linux-gnu/libgpg-error.so.0
-rw-r--r-- root/root            128944 2020-03-02 11:23 squashfs-root/usr/lib/x86_64-linux-gnu/libidn2.so.0
-rw-r--r-- root/root             35280 2020-02-13 13:20 squashfs-root/usr/lib/x86_64-linux-gnu/libip4tc.so.2
-rw-r--r-- root/root             67752 2020-02-19 17:32 squashfs-root/usr/lib/x86_64-linux-gnu/libjson-c.so.4
-rw-r--r-- root/root            108592 2020-04-18 18:53 squashfs-root/usr/lib/x86_64-linux-gnu/libkmod.so.2
-rw-r--r-- root/root            133464 2019-11-12 00:58 squashfs-root/usr/lib/x86_64-linux-gnu/liblz4.so.1
-rw-r--r-- root/root            162496 2019-08-24 16:28 squashfs-root/usr/lib/x86_64-linux-gnu/liblzma.so.5
-rw-r--r-- root/root           1325424 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/libm.so.6
-rw-r--r-- root/root            375360 2020-05-22 13:41 squashfs-root/usr/lib/x86_64-linux-gnu/libmount.so.1
-rw-r--r-- root/root             51696 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/libnss_files.so.2
-rw-r--r-- root/root             64176 2019-02-14 09:08 squashfs-root/usr/lib/x86_64-linux-gnu/libpam.so.0
-rw-r--r-- root/root            584360 2019-12-07 19:06 squashfs-root/usr/lib/x86_64-linux-gnu/libpcre2-8.so.0
-rw-r--r-- root/root            146912 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/libpthread.so.0
-rw-r--r-- root/root             39912 2020-05-11 03:04 squashfs-root/usr/lib/x86_64-linux-gnu/librt.so.1
-rw-r--r-- root/root            329960 2020-04-13 17:05 squashfs-root/usr/lib/x86_64-linux-gnu/libseccomp.so.2
-rw-r--r-- root/root            163520 2020-04-12 19:06 squashfs-root/usr/lib/x86_64-linux-gnu/libselinux.so.1
-rw-r--r-- root/root            174272 2020-05-17 20:28 squashfs-root/usr/lib/x86_64-linux-gnu/libudev.so.1
-rw-r--r-- root/root           1574952 2020-06-01 18:48 squashfs-root/usr/lib/x86_64-linux-gnu/libunistring.so.2
-rw-r--r-- root/root             30776 2020-05-22 13:41 squashfs-root/usr/lib/x86_64-linux-gnu/libuuid.so.1
drwxr-xr-x root/root                43 2020-06-08 19:29 squashfs-root/usr/lib64
-rwxr-xr-x root/root            169720 2020-05-11 03:04 squashfs-root/usr/lib64/ld-linux-x86-64.so.2
drwxr-xr-x root/root                26 2020-06-08 19:29 squashfs-root/var
lrwxrwxrwx root/root                 4 2019-05-11 13:42 squashfs-root/var/tmp -> /tmp
 ```

topimiettinen commented:
Doh, the `du` was of course referring to a compressed image. Here uncompressed:
```
# mount -o loop /usr/local/lib/rootimages/systemd-tmpfiles-all.squashfs /mnt
# du -hs /mnt/
26M     /mnt/
```

poettering commented:
26M is actually substantially less 236M... I wonder where precisely the discrepancy comes from?

That said, I was actually interested in the delta that tmpfiles pulls in relative to the minimal baseline distro image, i.e. the stuff that adding tmpfiles to a minimal baseline image actually adds, that it is responsible for. On Fedora at least glibc and libselinux are after all a given, and I'd assume included in *any* image built from it, so hardly something you can blame systemd-tmpfiles for. You can blame systemd-tmpfiles for pulling in stuff like the various compression libs and libcryptsetup and their deps however, and that's what I am interested in.

poettering commented:
> Well, fedora is not exactly "minimal"...

Well, custom built distros have it easy, they can just build everything they want and are done. it was my understanding this specific PR is about reducing the size of generic distribution images, i.e. an attempt to minimize the footprint a bit without losing all the flexibility the generic distros give you.

filbranden commented:
We do have a precedent with linking with static libshared in networkd and udev. For example, [networkd can be statically linked] with `-Dlink-networkd-shared=false`.

Yes, this one is somewhat different because it's building both at the same time.

I'd be happy to introduce something like `-Dlink-tmpfiles-shared=false` (and similarly for sysusers), that would allow us to accomplish the same thing in Fedora RPMs, even though in that case we would actually need to build the code twice, once with `-Dlink-tmpfiles-shared=true` (for the mainstream package) and another time with `-Dlink-tmpfiles-shared=false` (for the standalone one), but I guess that's workable...

Yet another option would be to control this at the libshared rule directly, exporting a new `-Dlibshared-static=true` (yeah I know haha) that would simply link *everything* to `libsystemd-shared.a` rather than the `.so`.

Does any of these options looks more palatable?

poettering commented:
It's the proliferation of compiling some bits "statically" and others not that concerns me. This really feels as if we should figure out the the details first and do our research properly. Otherwise we keep adding these hacks to one component here and one component there and then end up with a big pile of static components that cost us more than if we had done our homework first.

Note that I am not really against this PR at this point, all I am asking for is that we figure out things first, i.e. do our research before merging more stuff that makes stuff static.

Hence again: Those 236M, what do they consist of? What are those deps systemd is responsible for pulling in above what the minimal baseline container image already contains? It can't be so hard to figure out an answer to that question?

jamescassell commented:
> It's the proliferation of compiling some bits "statically" and others not that concerns me. This really feels as if we should figure out the the details first and do our research properly. Otherwise we keep adding these hacks to one component here and one component there and then end up with a big pile of static components that cost us more than if we had done our homework first.
> 
> Note that I am not really against this PR at this point, all I am asking for is that we figure out things first, i.e. do our research before merging more stuff that makes stuff static.
> 
> Hence again: Those 236M, what do they consist of? What are those deps systemd is responsible for pulling in above what the minimal baseline container image already contains? It can't be so hard to figure out an answer to that question?

Fedora 32, base of `glibc-minimal-langpack`, `coreutils-single`, and `systemd-libs`; versus adding `systemd` on top. 31 packages vs 97 packages; 45M installed vs 113M installed.

```
# root=/sr-empty packages='glibc-minimal-langpack coreutils-single systemd-libs' ; diff -uw <(yum --installroot=$root --releasever=/ --setopt=install_weak_deps=no install $packages --assumeno) <(yum --installroot=$root --releasever=/ --setopt=install_weak_deps=no install $packages systemd --assumeno) | grep ^'[-+]'
Operation aborted.
Operation aborted.
--- /dev/fd/63	2020-06-09 08:06:15.426386510 +0000
+++ /dev/fd/62	2020-06-09 08:06:15.427386515 +0000
-Last metadata expiration check: 1:56:44 ago on Tue Jun  9 06:09:31 2020.
+Waiting for process with pid 1497 to finish.
+Last metadata expiration check: 1:56:46 ago on Tue Jun  9 06:09:31 2020.
-================================================================================
- Package                   Arch      Version                   Repository  Size
-================================================================================
+======================================================================================
+ Package                       Arch    Version                          Repo      Size
+======================================================================================
+ systemd                       x86_64  245.6-2.fc32                     updates  4.2 M
+ acl                           x86_64  2.2.53-5.fc32                    fedora    72 k
+ alternatives                  x86_64  1.11-6.fc32                      fedora    35 k
+ audit-libs                    x86_64  3.0-0.19.20191104git1c2f876.fc32 fedora   115 k
+ bzip2-libs                    x86_64  1.0.8-2.fc32                     fedora    40 k
+ ca-certificates               noarch  2020.2.40-3.fc32                 fedora   358 k
+ cracklib                      x86_64  2.9.6-22.fc32                    fedora    83 k
+ crypto-policies               noarch  20200527-1.gitb234a47.fc32       updates   56 k
+ cryptsetup-libs               x86_64  2.3.3-1.fc32                     updates  470 k
+ dbus                          x86_64  1:1.12.18-1.fc32                 updates  9.7 k
+ dbus-broker                   x86_64  23-2.fc32                        updates  175 k
+ dbus-common                   noarch  1:1.12.18-1.fc32                 updates   16 k
+ device-mapper                 x86_64  1.02.171-1.fc32                  fedora   143 k
+ device-mapper-libs            x86_64  1.02.171-1.fc32                  fedora   180 k
+ elfutils-default-yama-scope   noarch  0.179-2.fc32                     updates   18 k
+ elfutils-libelf               x86_64  0.179-2.fc32                     updates  191 k
+ elfutils-libs                 x86_64  0.179-2.fc32                     updates  266 k
+ expat                         x86_64  2.2.8-2.fc32                     fedora   103 k
+ gawk                          x86_64  5.0.1-7.fc32                     fedora   1.2 M
+ gmp                           x86_64  1:6.1.2-13.fc32                  fedora   309 k
+ gnutls                        x86_64  3.6.14-1.fc32                    updates  993 k
+ gzip                          x86_64  1.10-2.fc32                      fedora   155 k
+ iptables-libs                 x86_64  1.8.4-7.fc32                     fedora   368 k
+ json-c                        x86_64  0.13.1-13.fc32                   updates   38 k
+ keyutils-libs                 x86_64  1.6-4.fc32                       fedora    31 k
+ kmod-libs                     x86_64  27-1.fc32                        fedora    64 k
+ krb5-libs                     x86_64  1.18-1.fc32                      fedora   756 k
+ libargon2                     x86_64  20171227-4.fc32                  fedora    30 k
+ libblkid                      x86_64  2.35.2-1.fc32                    updates  152 k
+ libcap-ng                     x86_64  0.7.10-2.fc32                    fedora    28 k
+ libcom_err                    x86_64  1.45.5-3.fc32                    fedora    28 k
+ libdb                         x86_64  5.3.28-40.fc32                   fedora   761 k
+ libfdisk                      x86_64  2.35.2-1.fc32                    updates  204 k
+ libffi                        x86_64  3.1-24.fc32                      fedora    33 k
+ libidn2                       x86_64  2.3.0-2.fc32                     fedora   104 k
+ libmnl                        x86_64  1.0.4-11.fc32                    fedora    28 k
+ libmount                      x86_64  2.35.2-1.fc32                    updates  179 k
+ libnetfilter_conntrack        x86_64  1.0.7-4.fc32                     fedora    62 k
+ libnfnetlink                  x86_64  1.0.1-17.fc32                    fedora    29 k
+ libnsl2                       x86_64  1.2.0-6.20180605git4a062cf.fc32  fedora    59 k
+ libpcap                       x86_64  14:1.9.1-3.fc32                  fedora   172 k
+ libpwquality                  x86_64  1.4.2-2.fc32                     fedora   101 k
+ libseccomp                    x86_64  2.4.2-3.fc32                     fedora    84 k
+ libsemanage                   x86_64  3.0-3.fc32                       fedora   123 k
+ libsigsegv                    x86_64  2.11-10.fc32                     fedora    26 k
+ libsmartcols                  x86_64  2.35.2-1.fc32                    updates  120 k
+ libtasn1                      x86_64  4.16.0-1.fc32                    fedora    73 k
+ libtirpc                      x86_64  1.2.6-0.fc32                     updates   97 k
+ libunistring                  x86_64  0.9.10-7.fc32                    fedora   512 k
+ libutempter                   x86_64  1.1.6-18.fc32                    fedora    26 k
+ libuuid                       x86_64  2.35.2-1.fc32                    updates   28 k
+ libverto                      x86_64  0.3.0-9.fc32                     fedora    22 k
+ libxcrypt                     x86_64  4.4.16-3.fc32                    updates  127 k
+ mpfr                          x86_64  4.0.2-4.fc32                     updates  256 k
+ nettle                        x86_64  3.5.1-5.fc32                     fedora   314 k
+ openssl-libs                  x86_64  1:1.1.1g-1.fc32                  updates  1.4 M
+ p11-kit                       x86_64  0.23.20-1.fc32                   fedora   347 k
+ p11-kit-trust                 x86_64  0.23.20-1.fc32                   fedora   142 k
+ pam                           x86_64  1.3.1-24.fc32                    fedora   666 k
+ qrencode-libs                 x86_64  4.0.2-5.fc32                     fedora    60 k
+ readline                      x86_64  8.0-4.fc32                       fedora   207 k
+ shadow-utils                  x86_64  2:4.8.1-2.fc32                   fedora   1.2 M
+ systemd-pam                   x86_64  245.6-2.fc32                     updates  299 k
+ systemd-rpm-macros            noarch  245.6-2.fc32                     updates   25 k
+ util-linux                    x86_64  2.35.2-1.fc32                    updates  2.7 M
+ zlib                          x86_64  1.2.11-21.fc32                   fedora    93 k
-================================================================================
-Install  31 Packages
+======================================================================================
+Install  97 Packages
-Total size: 13 M
-Total download size: 3.8 M
-Installed size: 45 M
+Total size: 34 M
+Total download size: 25 M
+Installed size: 113 M
```

poettering commented:
So, @jamescassell that is quite interesting, but still not the data that I think we should have before figuring out what to do: it adds in systemd.rpm as a whole. But I am just interested in the footprint that systemd-tmpfiles plus its immediate deps pull in. i.e. the systemd-tmpfiles binary, plus libsystemd-shared.so, plus its transient deps. But nothing of the other stuff in the big systemd.rpm.

LorbusChris commented:
https://src.fedoraproject.org/rpms/systemd/pull-request/25 would facilitate further exploration of that I would assume (i.e. splitting the `systemd` RPM into separate `systemd-shared`, `systemd-tmpfiles` and `systemd-sysusers` packages, without changing the actual build process)

poettering commented:
ah, that sounds interesting, but I only get "Sorry! This service is currently unavailable." if i click on that link

boucman commented:
for the cross compiling use-case (which is not the main use-case as I understand, but still worth mentionnig
These is the list of systemd tools buildroot needs to host-compile to properly install the target filesystem.

	systemd-machine-id-setup : statically generate a machine-id
	systemctl : run preset-all at the end of build
	udevadm : to statically generate the hwdb

there are also patch in review to add systemd-tmpfiles (statically generate files on read-only filesystems) systemd-sysuser (generate users for any package providing drop-ins) and journalctl (generate the catalogs)

anyway... there is a need to avoid compiling all systemd just for a couple of binaries, but this PR might not be the right place to discuss that...

filbranden commented:
I think @topimiettinen has the closest estimate, but that estimate has a `libsystemd-shared-v245.so` with ~7MB (which is consistent with one built from source, unstripped, etc.) while the one in the Fedora 32 RPM is closer to 2.8MB.

I just repeated that exercise by buiding the RPMs including this PR (and Fedora PR 27, but Fedora pagure is currently out, [status page](https://status.fedoraproject.org/) says there's a migration to a new datacenter going on.)

In any case, results for tmpfiles are:
 - With libsystemd-shared: 20,768 KiB
 - With standalone binaries (this PR): 4,748 KiB

The difference is 16,020 KiB, or 15.6 MiB.

Raw results are below. First for the build with libsystemd-shared:

```
$ find . -type f  -ls
 28181411    312 -rwxr-xr-x   1 filbranden filbranden   317984 Jun  9 13:41 ./lib64/libpthread.so.0
 28181406     72 -rwxr-xr-x   1 filbranden filbranden    70360 Jun  9 13:41 ./lib64/libpam.so.0
 28181400     40 -rwxr-xr-x   1 filbranden filbranden    37320 Jun  9 13:41 ./lib64/libip4tc.so.2
 28181427   1876 -rwxr-xr-x   1 filbranden filbranden  1918000 Jun  9 13:41 ./lib64/libm.so.6
 28181412     32 -rwxr-xr-x   1 filbranden filbranden    28904 Jun  9 13:41 ./lib64/libattr.so.1
 28181418     72 -rwxr-xr-x   1 filbranden filbranden    70800 Jun  9 13:41 ./lib64/libjson-c.so.4
 28181419    144 -rwxr-xr-x   1 filbranden filbranden   146992 Jun  9 13:41 ./lib64/libgpg-error.so.0
 28181420   1556 -rwxr-xr-x   1 filbranden filbranden  1592736 Jun  9 13:41 ./lib64/libunistring.so.2
 28181421    320 -rwxr-xr-x   1 filbranden filbranden   324184 Jun  9 13:41 ./lib64/libpcap.so.1
 28181392     44 -rwxr-xr-x   1 filbranden filbranden    41712 Jun  9 13:41 ./lib64/libacl.so.1
 28181399    136 -rwxr-xr-x   1 filbranden filbranden   135672 Jun  9 13:41 ./lib64/libidn2.so.0
 28181417     36 -rwxr-xr-x   1 filbranden filbranden    36800 Jun  9 13:41 ./lib64/libdl.so.2
 28181398   1168 -rwxr-xr-x   1 filbranden filbranden  1195808 Jun  9 13:41 ./lib64/libgcrypt.so.20
 28181414    376 -rwxr-xr-x   1 filbranden filbranden   383600 Jun  9 13:41 ./lib64/libdevmapper.so.1.02
 28181395     28 -rwxr-xr-x   1 filbranden filbranden    24992 Jun  9 13:41 ./lib64/libcap.so.2
 28181404   3024 -rwxr-xr-x   1 filbranden filbranden  3094744 Jun  9 13:41 ./lib64/libcrypto.so.1.1
 28181388   3112 -rwxr-xr-x   1 filbranden filbranden  3185792 Jun  9 13:41 ./lib64/libc.so.6
 28181407     76 -rwxr-xr-x   1 filbranden filbranden    75896 Jun  9 13:41 ./lib64/librt.so.1
 28181425    612 -rwxr-xr-x   1 filbranden filbranden   625352 Jun  9 13:41 ./lib64/libpcre2-8.so.0
 28181405   1240 -rwxr-xr-x   1 filbranden filbranden  1266504 Jun  9 13:41 ./lib64/libp11-kit.so.0
 28181422    104 -rwxr-xr-x   1 filbranden filbranden   103600 Jun  9 13:41 ./lib64/libz.so.1
 28181415    612 -rwxr-xr-x   1 filbranden filbranden   623608 Jun  9 13:41 ./lib64/libssl.so.1.1
 28181397    488 -rwxr-xr-x   1 filbranden filbranden   496856 Jun  9 13:41 ./lib64/libcryptsetup.so.12
 28181413     36 -rwxr-xr-x   1 filbranden filbranden    33216 Jun  9 13:41 ./lib64/libuuid.so.1
 28181409    176 -rwxr-xr-x   1 filbranden filbranden   176672 Jun  9 13:41 ./lib64/libselinux.so.1
 28181426    168 -rwxr-xr-x   1 filbranden filbranden   171128 Jun  9 13:41 ./lib64/libudev.so.1
 28181428     32 -rwxr-xr-x   1 filbranden filbranden    29168 Jun  9 13:41 ./lib64/libcap-ng.so.0
 28181394    336 -rwxr-xr-x   1 filbranden filbranden   343776 Jun  9 13:41 ./lib64/libblkid.so.1
 28181401    112 -rwxr-xr-x   1 filbranden filbranden   112408 Jun  9 13:41 ./lib64/libkmod.so.2
 28181396    204 -rwxr-xr-x   1 filbranden filbranden   206176 Jun  9 13:41 ./lib64/libcrypt.so.2
 28181393    108 -rwxr-xr-x   1 filbranden filbranden   108328 Jun  9 13:41 ./lib64/libgcc_s.so.1
 28181410    168 -rwxr-xr-x   1 filbranden filbranden   170184 Jun  9 13:41 ./lib64/liblzma.so.5
 28181403    388 -rwxr-xr-x   1 filbranden filbranden   396456 Jun  9 13:41 ./lib64/libmount.so.1
 28181408    312 -rwxr-xr-x   1 filbranden filbranden   316680 Jun  9 13:41 ./lib64/libseccomp.so.2
 28181423     44 -rwxr-xr-x   1 filbranden filbranden    41096 Jun  9 13:41 ./lib64/libffi.so.6
 28181402    132 -rwxr-xr-x   1 filbranden filbranden   132048 Jun  9 13:41 ./lib64/liblz4.so.1
 28181416     40 -rwxr-xr-x   1 filbranden filbranden    37584 Jun  9 13:41 ./lib64/libargon2.so.1
 28181424    128 -rwxr-xr-x   1 filbranden filbranden   127456 Jun  9 13:41 ./lib64/libaudit.so.1
 28181391   2780 -rwxr-xr-x   1 filbranden filbranden  2844272 Jun  9 13:41 ./usr/lib/systemd/libsystemd-shared-245.so
 28181386    100 -rwxr-xr-x   1 filbranden filbranden    99224 Jun  9 13:41 ./usr/bin/systemd-tmpfiles
```

And now for the standalone build:

```
 28181442    312 -rwxr-xr-x   1 filbranden filbranden   317984 Jun  9 13:42 ./lib64/libpthread.so.0
 28181439     32 -rwxr-xr-x   1 filbranden filbranden    28904 Jun  9 13:42 ./lib64/libattr.so.1
 28181435     44 -rwxr-xr-x   1 filbranden filbranden    41712 Jun  9 13:42 ./lib64/libacl.so.1
 28181441     36 -rwxr-xr-x   1 filbranden filbranden    36800 Jun  9 13:42 ./lib64/libdl.so.2
 28181436     28 -rwxr-xr-x   1 filbranden filbranden    24992 Jun  9 13:42 ./lib64/libcap.so.2
 28181434   3112 -rwxr-xr-x   1 filbranden filbranden  3185792 Jun  9 13:42 ./lib64/libc.so.6
 28181440    612 -rwxr-xr-x   1 filbranden filbranden   625352 Jun  9 13:42 ./lib64/libpcre2-8.so.0
 28181437    176 -rwxr-xr-x   1 filbranden filbranden   176672 Jun  9 13:42 ./lib64/libselinux.so.1
 28181438    108 -rwxr-xr-x   1 filbranden filbranden   108328 Jun  9 13:42 ./lib64/libgcc_s.so.1
 28181432    272 -rwxr-xr-x   1 filbranden filbranden   276264 Jun  9 13:42 ./usr/bin/systemd-tmpfiles.standalone
```

filbranden commented:
@poettering wrote:
> but still not the data that I think we should have before figuring out what to do: it adds in systemd.rpm as a whole. But I am just interested in the footprint that systemd-tmpfiles plus its immediate deps pull in. i.e. the systemd-tmpfiles binary, plus libsystemd-shared.so, plus its transient deps.

I think the problem with this rationale is that whether the footprint is large or small is on the eye of the beholder. 15.6MiB might look pretty small to you, but it can be considerably large for the container use case.

For instance, I just loaded container `registry.access.redhat.com/ubi8/ubi` and looked inside it:

```
[root@1d04f098c52b /]# du -ksx /
222436	/
```

(**But this container actually *already* includes all systemd packages, including libsystemd-shared and all its dependencies!**)

Even if you were considering a container even smaller than this one, 15.6MiB (out of 222,436KiB or less) would be 7.2% or more of the container size, and I'd say that's quite considerable. @LorbusChris if you have a more interesting target container size, that would be quite interesting too.

keszybz commented:
> Hence again: Those 236M, what do they consist of? What are those deps systemd is responsible for pulling in above what the minimal baseline container image already contains? It can't be so hard to figure out an answer to that question?

On the one hand, it's fairly easy to fire up `dnf --installroot` with some arbitrary package list. On the other hand, everyone's definition of "baseline" is different. So I don't think its possible to find one golden number here.

Instead, let's consider some specific cases: let's say that the container is using a semi-statically linked payload with a go or rust binary. Than the payload doesn't pull in any libraries above stuff like glibc which is installed anyway, and the payload is on the order of a few megabytes and negligible. Then the figures I quoted in https://github.com/systemd/systemd/pull/16061#issuecomment-639801490 are good approximate: we save about 236MB, half of the original image size.

If you want to do a deep dive, https://tiny.distro.builders/workloads.html is a very good page. Unfortunately, the graph of package dependencies is densely connected, so usually it's not so trivial to figure out what dependencies can be removed without doing a lot of work.

> i mean, it's fine doing such hacks downstream without the necessary research, if you want quick results.

The implementation is very clean. One of the reasons for the switch to meson was to be able to do such things in a reasonable way.

Dunno, I'm not sure what exactly we're discussing here. For me, it is fairly obvious that libsystemd-shared.so which pulls in a huge amount of libraries is the source of the issue. While having the shared library seems like a very good solution for "big" systems, it creates an issue for many "custom" systems which only need a subset of tools. And at that point, whether we save 20% or 30% or 50% is not particularly important: we **know** that the saving are going to be big, but that they are going to vary by case a lot too.

And yeah, this is not supposed to be the optimal and ideal solution. It's just a low-cost easy-to-test optimiation that works nicely for one particular use case, without inhibiting other solutions in any way.

filbranden commented:
@poettering @keszybz Ping?
